RecoLocalTracker/SiStripClusterizer/plugins/ChanLocsGPU.h

```diff
@@ -0,0 +1,146 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include <cuda_runtime.h>
+
+#include "HeterogeneousCore/CUDAUtilities/interface/device_unique_ptr.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/host_unique_ptr.h"
+#include "CUDADataFormats/SiStripCluster/interface/GPUtypes.h"
+
+class ChannelLocsGPU;
+
+template <template <typename> class T>
+class ChannelLocsBase {
```

Comment:
`ChannelLocsBase<T>` may hold data in GPU or in host memory, depending on `T` - but its members are declared as `__host__ __device__`, and only one of the two will be valid, depending on the specialization.

A somewhat cleaner implementation could be to declare the methods as `constexpr` rather than `__host__ __device`. The result will be the same, but at least it will not explicitly declare as `__device__` a host-only method or vice versa.

IMHO a better approach would be to declare `ChannelLocs` and `ChannelLocsGPU` independently, instead of inheriting from a common base. Is the common based used anywhere else, apart from convenience ?